KEP-9988: MultiKueue Global Quotas (#10145)
kubernetes-sigs/kueue: keps/9988-multikueue-global-quotas/README.md

1. For the manager quotas to serve their purposes discussed [above](#the-role-of-manager-quotas), they often must be **maintained in sync** with the worker quotas. \
   (The exact meaning of "in sync" may vary by use case; see [User Story 1](#story-1)). \
   Currently, MultiKueue does not help in this maintenance, making it entirely **user's responsibility**. \
   This is inconvenient and error-prone.

2. For a user to get the overview of actual resource availability and usage across the worker fleet, it is necessary to connect to all worker clusters and aggregate the information on their own.

3. For the MultiKueue controller on the manager cluster, having no awareness of workers' quotas (and their usage) **prevents more effective dispatching** choices. \
   Even though we generally don't want the manager to take over _the whole_ orchestrating work from the workers (as we intend to share it between both sides), some improvements are likely possible. For example, if a workload fits within the quota of worker A but not of worker B, there is no point in dispatching it to B before A. (At least, unless B could admit it through borrowing).

>    While designing specific improvements to dispatching is **out of scope** of this KEP, all this improvements require as the **first step** that the manager *knows* workers' quotas, which this KEP proposes. \

Reviewer: nit
```suggestion
   While designing specific improvements to dispatching is **out of scope** of this KEP, all these improvements require as the **first step** that the manager *knows* workers' quotas, which this KEP proposes. \
```
or "such"?

   Even when both those issues become closed, the fixes will be placed behind feature gates, making them still observable in the upcoming release.

3. **Borrowing on a worker**. This may raise the effective capacity of a worker CQ beyond its nominal quota. If the manager quota does not account for that, we risk under-utilization.

4. Other reasons _may_ also exist. For example - assuming hypothetically that a user wants to apply different preemption policies on different worker ClusterQueues (e.g. `Never` on one vs. `LowerPriority` on another) - "over-booking quota" on the manager would be the only way to have both policies honored without extra delays.

##### Potential reasons for decreasing manager quota

1. **Stray workloads on a worker**. When a LocalQueue `lq` points to a ClusterQueue `cq`, both on the manager and a worker, there are numerous ways to have a workload admitted to `cq` on the worker but on the manager:

   * the worker could be submitted directly to `lq` on the worker, "by-passing" MultiKueue;
   * the worker could be submitted to another LocalQueue `lq2` on the worker, where `lq2` also points to `cq` on the worker but is unrelated to MultiKueue;

>    * the worker could be submitted to another LocalQueue `lq2` on the worker, while `lq2` exists on the manager but points to another `cq2` with MultiKueue enabled.

Reviewer: Something reads off in this paragraph. I suppose you want to say "there are numerous ways to have a workload admitted to `cq` on a worker but not on the manager:":

Then you say "the workload could be submitted directly" - analogously in other points.

Author: Good points. Updated.

3. Let `R` = the set of all ClusterQueues on all workers to which at least one LocalQueue from `L` points.

(For simplicity, we've ignored resource borrowing between ClusterQueues, on either manager or worker side).

#### Treatment of ResourceFlavors

MultiKueue enforces **no relationships** between ResourceFlavors on the manager and worker sides. The flavors on both sides may have different names; they may even be differently grouped into ResourceGroups.

Even if the flavor names on both sides agree, having multiple flavors on the manager side may likely lead to **manager/worker flavor skew**, where a workload could be admitted on flavor A on the manager but on flavor B on a worker (as the manager's flavor assignment has typically no effect on the remotes). As long as such skews can happen, aggregating quotas per flavor would be meaningless.

As a result, multiple flavors on the manager side _should not be needed for most users_. Yet, they still may be needed in some special cases, for example:

* Using a flavor-aware custom MultiKueue dispatcher.
* Using TAS on the workers, with incompatible names of topology levels on different workers.

> Our proposal for now is to leave this cases unsupported. (Some other ideas are discussed below, see [here](#change-the-treatment-of-resourceflavors), [here](#support-quota-automation-for-multiple-manager-side-resourceflavors) and [here](#support-quota-automation-for-no-manager-side-resourceflavors-auto-create-one)).

Reviewer: ```suggestion
Our proposal for now is to leave these cases unsupported. (Some other ideas are discussed below, see [here](#change-the-treatment-of-resourceflavors), [here](#support-quota-automation-for-multiple-manager-side-resourceflavors) and [here](#support-quota-automation-for-no-manager-side-resourceflavors-auto-create-one)).
```
nit

3. Slowing down MultiKueue by adding more reconcilers and computations.

For these, we propose the following mitigations:

* Risk 1 is mitigated by a dedicated ClusterQueue condition, explaining the automation process.

* Risks 1 and 2 are mitigated by introducing a feature gate and the opt-in/opt-out API. \
  This also mitigates Risk 3, as long as we pay attention to skip computations which are not necessary per the ClusterQueue configuration.

* Risk 3 is considered low. If needed, it can be substantially mitigated by caching the relevant information.

## Design details

### API surface

The configuration for this feature will be added to the existing `MultiKueueConfig` struct. (For context, `MultiKueueConfig` objects are referenced from `AdmissionCheckSpec` as [`Parameters`](https://github.com/kubernetes-sigs/kueue/blob/24209c461b72fd6519581aca2234fb8f05dd1ce7/apis/kueue/v1beta2/admissioncheck_types.go#L60); this will allow controlling the feature independently for each ClusterQueue in the manager cluster).

```go
type MultiKueueConfigSpec struct {
   // ...

   // quotaAutomation specifies the automation behavior of ClusterQueue quotas 

>    // in the manager cluster.

Reviewer: Let's also add a comment, that the quotaAutomation is disabled when the structure is left unspecified.

Author: Adding a comment to clarify this is a great idea.

However, IIRC we actually intended it to be eventually _enabled_ by default?
So that default setting = feature gate state.

For now, I specified it that way. PTAL if it makes sense.

Reviewer: I see, thanks for clarifying that. the comments lgtm.

I'm also ok with the idea of enabling the sync by default, but we need to make sure this is well communicated in the release notes when graduating as a potentially breaking change for some users. Users who rely on custom sync logic (we don't know yet how common that would be) should set `quotaAutomation.mode=Disabled` prior to the upgrade to the version when the sync is enabled by default.

I would just add a remainder to communicate that clearly in release notes as another graduation point for Beta.

Author: Done.

This discrepancy may lead to temporary under-utilization of the fleet resources. For example, if there are 3 workers, each with 10 CPU capacity, and we're running a single workload, consuming 7 CPU, on `worker1` which gets disconnected, then the manager ClusterQueue will have its `nominalQuota` set to 20, with 7 of that already used, leaving just 13 CPU for new workloads on `worker2` and `worker3` which are both empty.

To cover for that, we could increase the manager-side quota by the sum of requests of all workloads which are admitted on the manager but their assigned worker clusters have remote clients which are currently `.connecting`. (Notably, all this info is available on the manager side, without querying the workers).

We postpone this because it adds some complexity, and the impact feels minor.

### Graduation Criteria

* **Alpha:**
  * The feature implemented behind the `MultiKueueManagerQuotaAutomation` feature gate, disabled by default.
  * Execution of the newly introduced API calls (local & remote) is guarded by feature gates.
  * Tests are implemented.

* **Beta:**
  * The feature gate is enabled by default.

>   * The feature has been succesfully used in production environment.

Reviewer: ```suggestion
  * The feature has been successfully used in a production environment.
```

  * [Possible follow-ups](#possible-follow-ups) have been reconsidered, based on any newly gathered feedback.
  * No warning signs on performance were seen (in particular, no news of any external automation of worker ClusterQueue quotas) - or, if any were, they have been addressed.
  * Issue [#10428](https://github.com/kubernetes-sigs/kueue/issues/10428) (see [Drawback #1](#drawbacks)) is diagnosed, its impact on quota automation is understood and any mitigations deemed necessary are implemented.

* **Stable:**
  * The feature (offered functionality, API surface and implementation) has stabilized, without raising concerns.

>   * The feature gate is removed.

Reviewer: Actually the practice is:
- Lock the feature gate for "stable"
- remove the FG in N+2 (N  - version for stable)

The reason is that a feature entering stable may happen still to have bugs, and sometimes we need to disable. If the FG is kept it is much less work. For this reason k8s introduces the concept of "locking" FG.

We only remove in N+2 to also give users more time to remove the setting from their env - Kueue would not startup if a dropped FG is specified in config.

Author: Updated.

Enabling quota automation will be subject to a **constraint** that the manager ClusterQueue has exactly one ResourceFlavor. This constraint will **not** be enforced via validation (e.g. a webhook) but by Kueue controllers code (similarly to [analogous existing AdmissionCheck-related constraints](https://github.com/kubernetes-sigs/kueue/blob/6163e91e5a62befbdd421097fe0ec38b37d406e0/pkg/cache/scheduler/clusterqueue.go#L284-L307), evaluated in the core ClusterQueue controller). A violation of this new constraint will **not** make the ClusterQueue Inactive (as is the case for the existing analogous constraints). In this case of a violation, the queue itself will be still operational, just not supportable by the quota automation feature. This will be communicated as an error of that feature, by its status Condition:

```yaml
type: MultiKueueManagerQuotaAutomation
status: False
reason: MultipleFlavors
message: MultiKueue manager quota automation does not support ClusterQueues with multiple ResourceFlavors.
lastTransitionTime: 2026-01-01T00:00:00Z
```

#### Cross-worker resource stats in Visibility API

> In the Kueue Visbility API, we will add a new [query](https://github.com/kubernetes-sigs/kueue/blob/3f0c4b2884fe5577d7a7ae4c3579a49718077be3/pkg/visibility/storage/storage.go#L26-L31), `clusterqueues/workerresourcestats`, allowing fetching two kinds of resource stats:

Reviewer 2: What kind of provider will you wire here? Just a simple rest.Getter?

Author: Closing here as the Visibility feature will be moved to another KEP.

* grouped by _worker-side_ flavors,
* tied to the _worker-side_ workload status (quota reserved / admitted),
* aggregating _worker-side_ quota reservations, that is: if a single manager workload has multiple remotes reserving quota (which is possible following [#8592](https://github.com/kubernetes-sigs/kueue/pull/8592), especially if worker clusters observe long-running AdmissionChecks), we'll add up those quota reservations on all the workers.

In comparison to the [existing manager-centric fields](https://github.com/kubernetes-sigs/kueue/blob/6163e91e5a62befbdd421097fe0ec38b37d406e0/apis/kueue/v1beta2/clusterqueue_types.go#L294-L309), we've also removed the limit of list length (set to [64 items](https://github.com/kubernetes-sigs/kueue/blob/6163e91e5a62befbdd421097fe0ec38b37d406e0/apis/kueue/v1beta2/clusterqueue_types.go#L299)). This is because the number of [related worker ClusterQueues](#defining-related-worker-clusterqueues) is theoretically unbounded, and each of those [could have different ResourceFlavor names](#treatment-of-resourceflavors).

> However, the new flavor utilization stats are **not** going to be just aggregates of the [existing ClusterQueueStatus fields](https://github.com/kubernetes-sigs/kueue/blob/6163e91e5a62befbdd421097fe0ec38b37d406e0/apis/kueue/v1beta2/clusterqueue_types.go#L294-L309) from all [related worker ClusterQueues](#defining-related-worker-clusterqueues). Doing so would include various kinds of "stray" workloads unrelated to the considered manager ClusterQueue (see [Potential reasons for decreasing manager quota](#potential-reasons-for-decreasing-manager-quota)). Instead, we will aggregate resource stats **only** for remotes **dispatched from** the given manager ClusterQueue.

Reviewer 2: Given this, we will calculate ReservedQuota (R) and UsedQuota (U) for manager remotes, but total quota (Q) in general?

If yes, there are two points that could cause confusion that I don't think are listed in the downsides:
1. Q may include resources not listed as available on the Manager CQ, as Workers may have other resources at their disposal.
2. Users may naturally expect that Q - (R + U) = Available Resources, which is incorrect, as R and U do not include resources held by other workloads on the worker (ones not managed by this Manager CQ). Shouldn't we include a "reserved by external workloads" stat to avoid this issue?

Author: Ad 1. The divergence between manager-side quotas and worker-side quotas is already discussed (as sth amplified by the mulitiplier).

Ad 2. Good point, I added this as a drawback.

Also, BTW, you reminded me to discuss borrowing (i.e. lack of support for it as a Drawback, and adding it as a Possible Follow-Up).

### MultiKueue Cache

Both features will be based on a new **MultiKueue Cache**, storing some information retrieved from the watched resources and fetched workloads.

The cache format will be tentatively as follows: \
(referring to the existing definitions of [`FlavorResourceQuantities`](https://github.com/kubernetes-sigs/kueue/blob/3f0c4b2884fe5577d7a7ae4c3579a49718077be3/pkg/resources/resource.go#L37) and [`LocalQueueReference`](https://github.com/kubernetes-sigs/kueue/blob/3f0c4b2884fe5577d7a7ae4c3579a49718077be3/pkg/util/queue/local_queue.go#L25-L26))

```go
type clusterReference string

type clusterQueueReference string

type workloadReference struct {
   Name string
   Namespace string
}

type FlavorsUtilization struct {
   Reserved FlavorResourceQuantities
   Admitted FlavorResourceQuantities 
}

type wlGroupResourceInfo struct {
   // This is not nil if and only if the workload has quota reserved locally.
   LocalClusterQueue *clusterQueueReference
   FlavorsUtilization FlavorsUtilization
}

type MultiKueueCache struct {

>    localLqToCqMap map[LocalQueueReference]clusterQueueReference

Reviewer 2: For manager LQs and CQs this mapping is already recorded in the queue cache (manager). Can't we utilize that?

Author: Hmm... it seems you're reviewing an old version?

### Goals

* Make the manager cluster aware of worker's quotas.
* Enable automated maintenance of manager quotas to keep it in sync with total worker quotas.
* Make the above automation configurable enough to fit, at least roughly, various use cases.

### Non-Goals

* Fine-tune manager quota automation to achieve the optimal scheduling throughput.
* Expose in the manager cluster a view of resource availability and usage across all connected worker clusters. (This is tracked in [#10105](https://github.com/kubernetes-sigs/kueue/issues/10105)).
* Improve MultiKueue dispatching.

## Proposal

We will add an **optional** feature of **auto-aggregating quotas** from the workers to the manager.

**When enabled** for a manager ClusterQueue `Q`, this feature will set its quota for a resource `R` according to the following recipe:

1. Determine the set of [related worker ClusterQueues](#defining-related-worker-clusterqueues), i.e. those which can serve the dispatched remotes of the workloads submitted to `Q`.

1. Compute the **sum** of quotas for `R` in **all** _related_ worker ClusterQueues, across **all** ResourceFlavors.

> 1. Multiply the above sum by a user-configurable **multiplier**. \

Reviewer 3: Will take a deeper dive today. I understand the motivations for the multiplier (ie preventing dispatcher from blocking admission in the example below `Divergence of quota reservation`), but initial reaction on skim is that it doesn't feel clean and is confusing. I do agree that we need something to sync quotas though.

Reviewer 3: I would separate the multiplier from this kep as a smaller kep after you get auto-aggregating quotas working/merged. To me they're separate. Combining them bloats this KEP.

Author: > _initial reaction on skim is that it doesn't feel clean and is confusing_

Agreed. I'm not happy about this multiplier. 

> _I would separate the multiplier from this kep. (...) To me they're separate. Combining them bloats this KEP._

Here I don't agree.
If we implement just aggregation without the multiplier, it would actually be a _degradation_ for almost everyone. (This is the bad news from the ["Divergence of quota reservation" story](https://github.com/olekzabl/kueue/blob/b0eb72c8999d5eb6d5a74a3862398125493278e9/keps/9988-multikueue-global-quotas/README.md#potential-reasons-for-increasing-manager-quota)).
So, on top of auto-aggregating, I think we *need* "Something" to make it useful (at least for a large part of users), even in the early version.

We've brainstormed (with @mwielgus and @mimowo ) a few options of what that "Something" could be, and we found nothing better than this.
I now expanded [one of Alternatives](https://github.com/olekzabl/kueue/blob/3fb8cd9a6c33f9fffbbd851229d85abbe7dfc6ee/keps/9988-multikueue-global-quotas/README.md#avoid-the-quota-automation-multiplier) discussing that; PTAL.

(Side note: I'm still warmly thinking of the [quota reservation overbooking multiplier](https://github.com/olekzabl/kueue/blob/3fb8cd9a6c33f9fffbbd851229d85abbe7dfc6ee/keps/9988-multikueue-global-quotas/README.md#introduce-a-multikueue-manager-clusterqueue-quota-reservation-overbooking-multiplier) instead of the proposed one, as I'm feeling this would at least make the quota values on the manager not confusing. However, this comes with its own disadvantages, and so far I've received feedback leaning against this).

Reviewer: > Agreed. I'm not happy about this multiplier.

Let's see if this is really expected by some MultiKueue practitioners: https://kubernetes.slack.com/archives/C032ZE66A2X/p1777280674564329 who already use MultiKueue -how they sync the quota currently in the manual way.

I'm not happy about that either, so I would consider this like a follow up if needed.

Reviewer 4: > it doesn't feel clean and is confusing

Was my first gut feeling too.

Reviewer: As we synced with @olekzabl and @mwielgus the multiplier is not part of the initial phase. We are planning to use it in the follow up, but in a slightly different form. I would suggest to drop it from the "Design Details" and for now mention only in the Notes section or Alternatives.

Author: Dropped, and mentioned as a new [Possible Follow-Up](https://github.com/olekzabl/kueue/blob/br35/keps/9988-multikueue-global-quotas/README.md#add-a-way-to-adjust-the-manager-side-quota).

   (This is to avoid exaggerated reduction of quotas for `Q` in case when a worker cluster disconnects temporarily, and the workloads running on it still hold quota on `Q`, until `workerLostTimeout` elapses).

Enabling this feature will be controlled by an API field (not just by a feature gate), as we want to permanently retain a way to opt out (see [User Story 3](#story-3)).

Enabling this feature will **require** that the manager ClusterQueue `Q` has **exactly one ResourceFlavor**. (See [Treatment of ResourceFlavors](#treatment-of-resourceflavors) for rationale).

### User stories

#### Story 1

As a Batch Admin, I want to maintain manager quotas  **reasonably synced** to total woker quotas, to achieve a desired gate-keeping behavior on the manager.

Here, the meaning of "reasonably synced" can vary by use case:

* The **baseline approach** (most intuitive, and currently recommended in MultiKueue documentation) is to keep the manager quota **equal to the sum** of worker quotas.

* Yet, there are several reasons for which the user may want to keep the manager quota **higher or lower** than the abovementioned sum. \

>   See [Factors influencing desired manager quota](#factors-influencing-desired-manager-quota) for more details.

Reviewer: AFAIK this is still a speculative part of the story. Let me check if this is actually used by the practitioners, on slack.

I would rather prefer to de-couple this speculative part of the story as a substory, or another story on its own.

cc @MichalZylinski, @mwielgus

Reviewer: asked the community on slack to guide the multiplier: https://kubernetes.slack.com/archives/C032ZE66A2X/p1777280674564329

Author: Please note that it's not framed as a User Story - but as "Potential reasons (...)" under "Notes/Constraints/Caveats".

And it's not just my happy imagination; it's mostly based on early-collected feedback.

> asked the community on slack to guide the multiplier:

Thank you for this, and I value the responses... yet, I think the number of responses there is too low to draw strong conclusions about "how much speculative" is that part.

Author: FTR the multiplier was degraded to a "Future work idea".
Resolving this thread.

### Goals

* Make the manager cluster aware of worker's quotas.
* Enable automated maintenance of manager quotas to keep it in sync with total worker quotas.
* Make the above automation configurable enough to fit, at least roughly, various use cases.

### Non-Goals

* Fine-tune manager quota automation to achieve the optimal scheduling throughput.
* Expose in the manager cluster a view of resource availability and usage across all connected worker clusters. (This is tracked in [#10105](https://github.com/kubernetes-sigs/kueue/issues/10105)).
* Improve MultiKueue dispatching.

## Proposal

We will add an **optional** feature of **auto-aggregating quotas** from the workers to the manager.

**When enabled** for a manager ClusterQueue `Q`, this feature will set its quota for a resource `R` according to the following recipe:

1. Determine the set of [related worker ClusterQueues](#defining-related-worker-clusterqueues), i.e. those which can serve the dispatched remotes of the workloads submitted to `Q`.

1. Compute the **sum** of quotas for `R` in **all** _related_ worker ClusterQueues, across **all** ResourceFlavors.

1. Multiply the above sum by a user-configurable **multiplier**. \

>    (This is intended to address several [reasons for adjusting the manager quota](#factors-influencing-desired-manager-quota), and thus cover most cases of [User Story 1](#story-1). Also, specifying a very large multiplier is technically a way to cover [User Story 2](#story-2)).

Reviewer: I think this is too detailed, and is repetition of the Design Details section. What we want to have here is that we introduce a new API in MultiKueueConfig, and the mechanism, to automatically set the quota in the manager "reasonably".

Author: The multiplier got removed, and I shortened Design Details following the other feedback bits. So no longer repetition, I think. PTAL.

1. Last, add the requests of all (manager-side) Workloads which are currently Admitted in `Q` but their remote cluster is currently unreachable. \

>    (This is to avoid exaggerated reduction of quotas for `Q` in case when a worker cluster disconnects temporarily, and the workloads running on it still hold quota on `Q`, until `workerLostTimeout` elapses).

Reviewer: This is too detailed for the "Proposal" section. It seems ok as a section like "Handling worker disconnection" in Design Details.

Author: This got obsolete. (Now discussed more roughly as a [Possible Follow-up](https://github.com/olekzabl/kueue/blob/br35/keps/9988-multikueue-global-quotas/README.md#account-for-temporarily-unavailable-workers)).

## Design details

### API surface

The configuration for this feature will be added to the existing `MultiKueueConfig` struct. (For context, `MultiKueueConfig` objects are referenced from `AdmissionCheckSpec` as [`Parameters`](https://github.com/kubernetes-sigs/kueue/blob/24209c461b72fd6519581aca2234fb8f05dd1ce7/apis/kueue/v1beta2/admissioncheck_types.go#L60); this will allow controlling the feature independently for each ClusterQueue in the manager cluster).

```go
type MultiKueueConfigSpec struct {
   // ...

   // quotaAutomation specifies whether (and how) the ClusterQueue quotas 
   // in the manager cluster should be automatically updated 
   // based on the total quota in all related worker ClusterQueues.
   QuotaAutomation *QuotaAutomation`json:"quotaAutomation,omitEmpty"`
}

type QuotaAutomation struct {
   // enabled specifies whether ClusterQueue quotas in the manager cluster
   // should be automatically set based on worker-side quotas.
   // The default value depends on the feature gate MultiKueueManagerQuotaAutomation.

>    Enabled *bool `json:"enabled,omitEmpty"`

Reviewer: I can certainly understand the motivation for the "enabled: true" field like this, but currently such fields are not recommended in the core k8s https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventions.md#primitive-types. 

Some ideas that come to me:

```yaml
spec:
   quotaSync: Manual / Automated// enum
   quotaAutomation:
      // params specific for the quota automation
```
or:
```yaml
spec:
   quotaAutomation: Disabled / Enabled // enum
   quotaAutomationConfig:
      // params specific for the quota automation
```

Author: I take the point that "booleans are not recommended"; thank you for this.

Yet, I don't feel why split the management of a single feature to 2 top-level fields?
(Just to avoid nesting?)

In my head, it's still rather sth like
```yaml
spec:
  quotaAutomation:
    mode: Enabled / Disabled // enum
    param1: ...
    param2: ...
```

So for now I made sth like this. PTAL.

Author: FTR this was split back to 2 top-level fields based on later feedback.
Resolving this thread.

   // quotaMultiplier will be applied on top of the total worker-side quotas
   // in order to define the manager-side quota to be automatically set.
   // This value is ignored if the quota automation feature is disabled.

>    // Defaults to 3.

Reviewer: I think the default should be 1, the other +/- is still a speculative user story in my mind. I would also like to add some validation that only +/- 50% is possible at max.

Author: > I think the default should be 1

The whole point of having this multiplier proposed in this KEP (and not as some possible follow-up) is that we believe the value 1 to be _wrong_ for most users.

> I would also like to add some validation that only +/- 50% is possible at max.

3 was suggested by @mwielgus who claimed it should be roughly that large.

Reviewer: > The whole point of having this multiplier proposed in this KEP (and not as some possible follow-up) is that we believe the value 1 to be wrong for most users.

In my understanding we introduce the multiplier to support some speculative use cases, which may be or not there. This does not imply the value "3" is generally better than "1" for most users.

The claim that "3" is better than "1" and counter-intuitive to me, and thus extraordinary. Extraordinary claims require extraordinary evidence, so I would like to collect such feedback from users. We already have some adoption of MultiKueue, so let's explore such channels, to learn how they currently establish the quotas on the manager cluster manually. I started the discussion on slack: https://kubernetes.slack.com/archives/C032ZE66A2X/p1777280674564329

Reviewer: Also, note that our documentation is quite clear recommending our users 1: https://kueue.sigs.k8s.io/docs/concepts/multikueue/#manager-cluster

So I think it is more sensible for our defaults to correspond to the recommendation. Still I'm ok to introduce the `multiplier` to work within some reasonable renge: [50% - 200$], 300% seems like a lot. We can always relax validation if someone requests. We will then also get a clear signal that someone needs high value. If we open it, then we may never know what is really used.

The quota set for the flavors of a ClusterQueue determines how many jobs are eligible for dispatching at a given time. Ideally, the quota in the manager cluster should equal the total quota available in all worker clusters:

If the manager’s quota is significantly lower, worker clusters may remain underutilized.
If the manager’s quota is significantly higher, it may dispatch and monitor workloads that are unlikely to be admitted in the worker clusters.

Author: This got obsolete.

>    QuotaMultiplier *float32 `json:"quotaMultiplier,omitEmpty"`

Reviewer: k8s conventions on API discourage float fields: https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventions.md#primitive-types

I would rather replace it (if we need the multiplier field at all) with 

```yaml
quotaMultiplierPercentValue: 130
quotaMultiplierMillisValue: 1300
```
both would use ints

Author: Thank you for flagging.

How about `resource.Quantity` then? I'm feeling it'd be clearer to see
```
multiplier: 3
multiplier: 1500m
```
than
```
multiplierPercentValue: 300
multiplierPercentValue: 150
```

Reviewer: This is a possible alternative, but `resource.Quantity` is typically used for resource / quota, not dimentionless numbers. 

For example with resource.Quantity we can say "0.1Mi" which is technically correct but would be misleading to specify "102.4" (iiuc). 

If we just need "percent" resolution then we have some examples for using int32 with the "percent" / "percentage". The best known would be in the core k8s for P&F, like: https://github.com/kubernetes/kubernetes/blob/75d51c440770f96b9725efea4145ba50fc6aef4b/pkg/apis/flowcontrol/types.go#L483-L492, https://github.com/kubernetes/kubernetes/blob/75d51c440770f96b9725efea4145ba50fc6aef4b/pkg/apis/flowcontrol/types.go#L459

However, I understand these use cases are more about small "percent" rather than generic "multiplier" which could be >100 percent.

Reviewer: As discussed https://github.com/kubernetes-sigs/kueue/pull/10145#discussion_r3167389033

Author: This got obsolete.

When quota automation is not requested, the condition will look like this:

```yaml
type: MultiKueueManagerQuotaAutomation
status: False
reason: NotRequested
message: MultiKueue manager quota automation has not been requested.
lastTransitionTime: 2026-01-01T00:00:00Z
```

Enabling quota automation will be subject to a **constraint** that the manager ClusterQueue has exactly one ResourceFlavor. This constraint will **not** be enforced via validation (e.g. a webhook) but by Kueue controllers code (similarly to [analogous existing AdmissionCheck-related constraints](https://github.com/kubernetes-sigs/kueue/blob/6163e91e5a62befbdd421097fe0ec38b37d406e0/pkg/cache/scheduler/clusterqueue.go#L284-L307), evaluated in the core ClusterQueue controller). A violation of this new constraint will **not** make the ClusterQueue Inactive (as is the case for the existing analogous constraints). In this case of a violation, the queue itself will be still operational, just not supportable by the quota automation feature. This will be communicated as an error of that feature, by its status Condition:

```yaml
type: MultiKueueManagerQuotaAutomation
status: False

> reason: MultipleFlavors

Reviewer: I would rather call it more generically like "UnsupportedConfiguration" I don't think MultipleFlavors are intrinsically a problem, because:
1. we may need to disable also the automation in case of Cohorts (for now)
2. I don't think MultipleFlavors is intrinsically problematic. I think we may atually need two flavors: TAS/non-TAS, or even more flavors per TAS type.

Author: > 1. we may need to disable also the automation in case of Cohorts (for now)

If we decide so, it will sound like a good case for another `reason`.
Why not make `reason` (roughly) informative for the users?

> 2. I don't think MultipleFlavors is intrinsically problematic. 

I do claim it is intrinsically problematic.
See the [discussion in KEP](https://github.com/olekzabl/kueue/blob/3fb8cd9a6c33f9fffbbd851229d85abbe7dfc6ee/keps/9988-multikueue-global-quotas/README.md#support-quota-automation-for-multiple-manager-side-resourceflavors) and [in another comment](https://github.com/kubernetes-sigs/kueue/pull/10145#discussion_r3146579187).

> I think we may atually need two flavors: TAS/non-TAS

Such a need I don't see.
If anyone would need 1 TAS flavor and 1 non-TAS flavor on the manager, the TAS flavor should cover well for both (of course, with quotas summed up). Because it will be capable of admitting also the non-TAS workloads.

> or even more flavors per TAS type.

See [the other comment thread](https://github.com/kubernetes-sigs/kueue/pull/10145#discussion_r3146579187) specifically on this.

Reviewer: > If we decide so, it will sound like a good case for another reason.
> Why not make reason (roughly) informative for the users?

If this is to be informative to users, then the `message` field should be sufficient.  The `reason` field is typically for automation. 

I'm not sure if automation needs to really read this "Reason". I think this is mostly something that the user will read only anyway.

> I do claim it is intrinsically problematic.
> See the discussion in KEP and in another comment.

To show there are use cases: I've seen deployments, production when users would send to the same CQ GPU and CPU workloads, assuming they would land only on one of the flavors. For example see this issue: https://github.com/kubernetes-sigs/kueue/issues/8095

I think having two separate quotas for GPU and CPU at the manager level is reasonable for such a deployment. Otherwise if GPU is saturated already then the manager will keep admitting GPU workloads just because it thinks there is still more of the "virtual CPU + GPU" resource. As a result the GPU workload may block the CPU processing even more. 

I totally agree supporting multiple flavors is complex, and opens difficult questions. And I'm totally ok to start with a single flavor for Alpha. 

I just want the "reason" to sound less strong about "MultipleFlavors" are the problem, but "We currently do not support the configuration with multiple flavors".

Author: I renamed the reason.

As for the rest of this discussion - we already followed up deeper in [another comment thread](https://github.com/kubernetes-sigs/kueue/pull/10145#discussion_r3146579187).

message: MultiKueue manager quota automation does not support ClusterQueues with multiple ResourceFlavors.
lastTransitionTime: 2026-01-01T00:00:00Z
```

> ### Guiding principles for the implementation

Reviewer: I think these sections here (and "MultiKueue ClusterQueue Reconciler") are too detailed for KEP. 

Yes, it is ok to include some of the implementation overview in the KEP, but it would rather be **declarative** listing the responsibilities of various components, rather than **procedural** description of their work.

The reason is simple: KEP maintanance overhead, becasue:
- the points describing the Reconciler work are likely to change as we extend the KEP with new features. 
- some implementation details are only figured out / discovered when working with the code.

I fully agree it is good to thing e2e of where code needs to be changed, but for that usually prototypes are a good idea, or external design docs linked.

Author: I would understand this point for "MK CQ Reconciler". 
(Though even there, I'd still argue that I saw details of analogous level in recent KEPs I reviewed, and I considered those valuable, even without a prototype).

However, in this particular section ("Guiding principles"), I disagree more fundamentally :)
I claim this section **is** declarative (not "procedural").
Listing maybe not "responsibilities" but "principles" on which we want to base.
And my goal is to have this aligned on even before prototyping (as otherwise prototyping could be just wasted effort).

Reviewer: > I would understand this point for "MK CQ Reconciler".

Yes, it is the "MK CQ Reconciler" section that made me make post this comment.

Once I decided to make the comment I returned back to place it on "Guiding principles for the implementation", because it also feels too detailed, but still "on the edge" (indeed, for a different reason, see below).

> And my goal is to have this aligned on even before prototyping (as otherwise prototyping could be just wasted effort).

I agree with the principles, but I don't get from them more than: "accept suboptimal performance practices  like infrequent requests on remote clients, until proven that we need to optimize. Once proven use caching.".  This is captured by generic enginesesing principles like KISS, and avoid premature optimization. 

Yes, they add a bit more details which kind of optimizations are deferred for later, but this still feels like generic "Contributor guidelines".

If you feel strongly this should be stated explicitly in this KEP, then ok. I'm just trying to minimize the unnecessary parts so that me (and other reviewers) can put all our attention into the relevant parts of the KEP.

Reviewer 3: Yeah... there's a lot detail here. I'd remove these sections:
- Generously tolerate API calls for resources which typically don't exist in large numbers.
- Still, to keep up with Kueue code standards, we will
- MultiKueue ClusterQueue Reconciler
- MultiKueue Cluster Reconciler

Tbh you can shave down some of the testing section. I like when people keep testing section to testing stuff that's not obvious. Because its not like you won't be involved in the implementation to see this vision through as either implementer or reviewer for the obvious bits.

If you really want principals, pull it out and generalize Kueue developer principals in the docs or something. Even better make a new reviewer or implementation skill in the new AGENTS.md file (shameless plug 😂) Or a new linting rule if thats possible for cicd. 

More seriously, these all seem generalizable and principals people tend to follow for controller development. There's already a lot to process for this doc.

Author: OK, so it seems I got out of sync in at least 2 aspects ("what is expected from a KEP" & "what is obvious to a Kueue developer"). Thank you for the hints; I've followed _in that direction_.

Still, out of the "guiding principles", I'm keeping [this](https://github.com/olekzabl/kueue/blob/br35/keps/9988-multikueue-global-quotas/README.md#a-note-on-client-usage). I'm still feeling there's some small bit here that isn't "just generic to any Kueue development"; after all, not every feature can afford happily sending client calls of any type. So I'm trying to distill what's special in this one which allows such "relaxed" approach.

And I hope this time it's digestably short :)

Reconciling a manager ClusterQueue `Q` will proceed as follows:

1. Fetch `Q` from the API client.

2. Find the MultiKueue AdmissionCheck for `Q` (using an unfiltered local API `.List()` call). \
   Then, fetch the MultiKueueConfig for `Q` from the API client. \
   If any of this fails, clear the `MultiKueueManagerQuotaAutomation` Condition (if present) and exit.

3. Determine whether quota automation has been requested and can be supported for `Q` (based on the MultiKueueConfig, the feature gate, and the number of ResourceFlavors). \
   If not, set the `MultiKueueManagerQuotaAutomation` Condition accordingly (to `False`) and exit.

5. Find all manager LocalQueues connected to `Q` (using a filtered local API `.List()` call).

6. Fetch all worker LocalQueues of those names (using an unfiltered remote API `.List()` call).

7. Fetch all worker ClusterQueues connected to those LocalQueues (using a single unfilterd remote API `.List()` call).

> 8. Determine the new quota values for `Q`.

Reviewer 2: To confirm - is this where QuotaMultiplier will be used? The collected quota will be adjusted and set on the Manager?

If yes - please add an explicit mention here, as the QuotaMultiplier param is only mentioned in setup and expansion-proposal and it is not clear where exacrly in the core proposal it is currently being utilized.

Author: PTAL how clear it's now.

   1. Fetch the corresponding manager LocalQueue from the API client, and check its ClusterQueue.
   2. Trigger the [MultiKueue ClusterQueue Reconciler](#multikueue-clusterqueue-reconciler) for that ClusterQueue.

2. For (remote) ClusterQueue events (creations, deletions, changes of `.Spec.ResourceGroups`), where the update handler function will, for a ClusterQueue `Q` on worker `W`:
   1. Find all LocalQueues on `W` pointing to `Q` (using a filtered remote `.List` API call).
   2. Find all manager ClusterQueues connected with manager LocalQueues of the same names (using a single unfiltered local `.List` API call).
   3. Trigger the [MultiKueue ClusterQueue Reconciler](#multikueue-clusterqueue-reconciler) for all those ClusterQueues.

### MultiKueue Cache

A new **MultiKueue Cache** will store the minimal information on the watched remote workloads that is needed to avoid bulk-fetching them from worker clusters. This is needed only for the "resource visibility" part, and will be maintained only when the `MultiKueueWorkerResourceStats` feature gate is enabled.

The cache format will be tentatively as follows:

```go
type workloadReference struct {
   Name string
   Namespace string

> }

Reviewer 2: Why not use `workload.Reference`?

Author: Good idea, thanks.

   Just like Drawback #2, this may be addressable by the [quota reservation overbooking](#introduce-a-multikueue-manager-clusterqueue-quota-reservation-overbooking-multiplier) idea, if used _together_ with the nominal quota multiplier as in the current proposal. Yet, the overall complexity of such solution feels unacceptable for now. We may consider it later depending on the feedback.

1. In the quota automation feature, the proposed solution fails to react to some types of detectable worker connection issues. \
   See [#10428](https://github.com/kubernetes-sigs/kueue/issues/10428) which describes this problem in more detail, and speculates that it may already affect existing MultiKueue reconcilers (e.g. the one for Workloads).

   If this turns out problematic, it should be fixable by extending the [MultiKueue Cache](#multikueue-cache) with the "manager ClusterQueue <-> MultiKueueCluster" relationship.

> 1. The proposed cross-worker resource stats fall short of giving a full and clear picture of resource availability and usage on the workers. \

Reviewer 2: I'd consider the option of adding the "external usage" statistics as one of the "possible follow ups" and linking that section of the KEP here as a "mitigation" option.

This is **shelved** for now due to the added complexity, both in implementation and specific desired behavior. Example questions:

* Should we simply increase the total quota by the `borrowingLimit`s on the workers?
* If borrowing does not allow preemption, should we rather define "borrowable quota" as "the minimum of borrowing limit and current unused capacity on the lending ClusterQueues"?
* Should we expose an analogue of [the `Borrowed` field](https://github.com/kubernetes-sigs/kueue/blob/6163e91e5a62befbdd421097fe0ec38b37d406e0/apis/kueue/v1beta2/clusterqueue_types.go#L359) in [cross-worker resource stats](#cross-worker-resource-stats-in-visibility-api-1)? Then, if a worker-side ClusterQueue used borrowing to host MultiKueue workloads from a manager ClusterQueue `Q` as well as various ["stray" Workloads](#potential-reasons-for-decreasing-manager-quota), how should we report that?

### Graduation Criteria

This design is covered by two feature gates:

> * `MultiKueueMaangerQuotaAutomation` - for the [quota automation part](#manager-quota-automation);

Reviewer 5: ### 
```suggestion
* `MultiKueueManagerQuotaAutomation` - for the [quota automation part](#manager-quota-automation);
```

3. For the MultiKueue controller on the manager cluster, having no awareness of workers' quotas (and their usage) makes dispatching choices less effective. \
   Even though we generally don't want the manager to take over _the whole_ orchestrating work from the workers (as we intend to share it between both sides), some improvements are likely possible. For example, if a workload fits within the quota of worker A but not of worker B, there is no point in dispatching it to B before A. (At least, unless B could admit it through borrowing).

   While designing specific improvements to dispatching is **out of scope** of this KEP, they all **depend** on building a mechanism allowing for manager awareness of workers' quotas. \
   Such a mechanism can be naturally implemented as an extension of the [MultiKueue Cache](#multikueue-cache) which this KEP proposes.

### Goals

* Make the manager cluster aware of worker's quotas, at real time.
* Expose in the manager cluster a view of resource availability and usage across all connected worker clusters.

> * Enable automated maintenance of manager quotas to keep it in sync with total worker quotas.

Reviewer 5: I briefly mentioned this offline, but I want to flag it here explicitly.

How do we want to treat transient issues? Consider a case where we have N workers, running at full capacity. Suddenly one of the workers goes down. So we are dealing with an oversubscription, since IIUC we'll update the automated quota on the manager to a lower value. 

Workloads have a `workerLostTimeout` and they will be unadmitted when it elapses, but in the meantime we could potentially encounter a disruption, no? For example a new batch of workloads come in. Since we are oversubscribed, we could end up preempting too many "healthy" workloads to free up quota.

Regardless of whether the proposed solution to this is to have to multiplier set up, I feel that this should be more explicitly considered in this KEP. WDYT?

Author: Good question.
My first impression was that this problem doesn't exist, but indeed, it does.

My response is adding step 4 to [Proposal](https://github.com/olekzabl/kueue/blob/93ea2aeaee5867d149defc69e0a0dc03c697340b/keps/9988-multikueue-global-quotas/README.md#proposal) and correspondingly steps 7-8 to the [MK CQ Reconciler recipe](https://github.com/olekzabl/kueue/blob/93ea2aeaee5867d149defc69e0a0dc03c697340b/keps/9988-multikueue-global-quotas/README.md#multikueue-clusterqueue-reconciler). PTAL.

I think it may be not _perfect_ as it can still lead to some evictions which might _turn out_ to be undesired when the worker comes back.
OTOH, I think the proposed formula makes sense. When a worker gets disconnected, we give the grace period (`workerLostTimeout`) to whatever is already running there - but we don't dispatch new stuff to it. So, if we ask what's the total mass of workloads that can be "admitted or dispatched" right now - it's just that.

Reviewer 5: I think this is an improvement and the right approach, as it preserves the existing "grace period" semantics for the workloads. That being said, maybe you could flag this edge-case in the KEP explicitly?

Author: Update: we decided that it's too "advanced" for the alpha scope. Degraded to a [Possible Follow-up](https://github.com/olekzabl/kueue/blob/br35/keps/9988-multikueue-global-quotas/README.md#account-for-temporarily-unavailable-workers).

reason: NotRequested
message: MultiKueue manager quota automation has not been requested.
lastTransitionTime: 2026-01-01T00:00:00Z
```

Enabling quota automation will be subject to a **constraint** that the manager ClusterQueue has exactly one ResourceFlavor. This constraint will **not** be enforced via validation (e.g. a webhook) but by Kueue controllers code (similarly to [analogous existing AdmissionCheck-related constraints](https://github.com/kubernetes-sigs/kueue/blob/6163e91e5a62befbdd421097fe0ec38b37d406e0/pkg/cache/scheduler/clusterqueue.go#L284-L307), evaluated in the core ClusterQueue controller). A violation of this new constraint will **not** make the ClusterQueue Inactive (as is the case for the existing analogous constraints). In this case of a violation, the queue itself will be still operational, just not supportable by the quota automation feature. This will be communicated as an error of that feature, by its status Condition:

```yaml
type: MultiKueueManagerQuotaAutomation
status: False
reason: MultipleFlavors
message: MultiKueue manager quota automation does not support ClusterQueues with multiple ResourceFlavors.
lastTransitionTime: 2026-01-01T00:00:00Z
```

> #### Cross-worker resource stats in Visibility API

Reviewer: I prefer to keep it out of scope for 0.18, because the code freeze is only in a month or so. I think we can move it to a follow up KEP, or update to the KEP. We can for now keep it under the Alternatives section.

Reviewer: > We can for now keep it under the Alternatives section.

Or just move it out as "non-goal" as content for another KEP. I think this makes sense as this part has big overlap with other KEPs coming.

Moreover I think the new interface ReconcilerWithFollowerObserver is a good idea on its own we could use in a preparatory refactoring already.

Author: I removed the visibility part from this KEP.

As for `ReconcilerWithFollowerObserver`, this belonged to the visibility scope, so we can come back to this in another KEP which I'll post soon.

   * The above approach is **insufficient** in our case, as [MultiKueue Cache](#multikueue-cache) depends on _remote_ API Workloads which we need to fetch in the `.Reconcile()` method. This logic is too heavy to be moved to predicates.

     Therefore, for our case, we propose a **variation** of [`leader_aware_reconciler`](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/leader_aware_reconciler.go), which would accept a controller implementing a richer interface

     ```golang
     interface ReconcilerWithFollowerObserver {
       TypedReconciler[Request]
       Observe(context.Context, request) (Result, error)
     }
     ```

     and, when about to reconcile an event, delegate to `.Reconcile()` or `.Observe()`, depending on the current leader/follower role.

>      Then, we'll make the MultiKueue Workload Reconciler wrapped with that, with the `.Observe()` method responsible only for updating [MultiKueue Cache](#multikueue-cache). (Internally, there will be some code sharing to avoid duplication).

Reviewer 2: How will this wrapper handle the `func (w *wlReconciler) setupWithManager(mgr ctrl.Manager) error` method? If it relies on the ReconcilerWithFollowerObserver interface, we will not have access to:
* `w.client`
* `w.clusters.wlUpdateCh`
* `w.roleTracker`
* `w.eventsBatchPeriod`
which are used in the method to set up the Controller.

How will this interface handle other Reconcilers we would like to update this way in the future (e.g. Cluster Reconciler) - each of them faces similar issues with the setupWithManager method.

Author: I think we can re-use the solution for the existing `leader_aware_reconciler` wrapper. Namely, the wrapper is called around an already-set-up "delegate". This happens here:

https://github.com/kubernetes-sigs/kueue/blob/7ec2aec50bc3bae7469f6816937555fd0e6f832f/pkg/controller/core/leader_aware_reconciler.go#L45-L58

and is used e.g. [here](https://github.com/kubernetes-sigs/kueue/blob/7ec2aec50bc3bae7469f6816937555fd0e6f832f/pkg/controller/core/clusterqueue_controller.go#L606) - so, strictly speaking, the wrapper is called around a delegate reconciler (`r`) that _is just being set up_ with the manager.

     ```golang
     interface ReconcilerWithFollowerObserver {
       TypedReconciler[Request]
       Observe(context.Context, request) (Result, error)
     }
     ```

     and, when about to reconcile an event, delegate to `.Reconcile()` or `.Observe()`, depending on the current leader/follower role.

     Then, we'll make the MultiKueue Workload Reconciler wrapped with that, with the `.Observe()` method responsible only for updating [MultiKueue Cache](#multikueue-cache). (Internally, there will be some code sharing to avoid duplication).

2. In [#10553](https://github.com/kubernetes-sigs/kueue/issues/10553), it's been proposed to separate Visibility API into a separate Pod / Deployment. This will clearly make it impossible to pass it any cached data from the "proper" Kueue binary.

   The proposed solution for this (initially agreed on in [#10553](https://github.com/kubernetes-sigs/kueue/issues/10553)) is to equip the Visibility Server binary with a sufficient set of controllers (preferably read-only ones) to provide the necessary information, partially duplicating some of the logic of the already existing Kueue controllers.

>    This can be implemented with a very similar mechanism as in the previous point: the Visibility Server simply needs to run the `.Observe()` method (to populate the cache) but not `.Reconcile()` (no need to update the Worlkoads). Thus, we can equip it with an **adjusted version** of the MultiKueue Workload Reconciler, which delegates `.Reconcile()` to execute just `.Observe()` even on the leader replica.

Reviewer 5: ```suggestion
   This can be implemented with a very similar mechanism as in the previous point: the Visibility Server simply needs to run the `.Observe()` method (to populate the cache) but not `.Reconcile()` (no need to update the Workloads). Thus, we can equip it with an **adjusted version** of the MultiKueue Workload Reconciler, which delegates `.Reconcile()` to execute just `.Observe()` even on the leader replica.
```
